pkg/controllers/organization/organization_controller.go

```
@@ -257,21 +257,40 @@ func (r *OrganizationReconciler) checkSCIMAPIAvailability(ctx context.Context, o
+	var authType *scim.AuthType
+	var basicAuthUser, basicAuthPw, bearerToken string
+	var err error
+	if scimConfig.BasicAuthUser.Secret != nil && scimConfig.BasicAuthPw.Secret != nil { // Conditions to avoid panic
+		basicAuthUser, err = clientutil.GetSecretKeyFromSecretKeyReference(ctx, r.Client, namespace, *scimConfig.BasicAuthUser.Secret)
+		if err != nil {
+			return greenhousesapv1alpha1.FalseCondition(greenhousesapv1alpha1.SCIMAPIAvailableCondition, greenhousesapv1alpha1.SecretNotFoundReason, "BasicAuthUser missing")
+		}
+		basicAuthPw, err = clientutil.GetSecretKeyFromSecretKeyReference(ctx, r.Client, namespace, *scimConfig.BasicAuthPw.Secret)
+		if err != nil {
+			return greenhousesapv1alpha1.FalseCondition(greenhousesapv1alpha1.SCIMAPIAvailableCondition, greenhousesapv1alpha1.SecretNotFoundReason, "BasicAuthPw missing")
+		}
+		newAuthType := scim.Basic
+		authType = &newAuthType
 	}
-	basicAuthPw, err := clientutil.GetSecretKeyFromSecretKeyReference(ctx, r.Client, namespace, *scimConfig.BasicAuthPw.Secret)
-	if err != nil {
-		return greenhousesapv1alpha1.FalseCondition(greenhousesapv1alpha1.SCIMAPIAvailableCondition, greenhousesapv1alpha1.SecretNotFoundReason, "BasicAuthPw missing")
+	if scimConfig.BearerToken.Secret != nil {
+		bearerToken, err = clientutil.GetSecretKeyFromSecretKeyReference(ctx, r.Client, namespace, *scimConfig.BearerToken.Secret)
+		if err != nil {
+			return greenhousesapv1alpha1.FalseCondition(greenhousesapv1alpha1.SCIMAPIAvailableCondition, greenhousesapv1alpha1.SecretNotFoundReason, "BearerToken missing")
+		}
+		newAuthType := scim.BearerToken
+		authType = &newAuthType
+	}
+	if authType == nil {
+		return greenhousesapv1alpha1.FalseCondition(greenhousesapv1alpha1.SCIMAPIAvailableCondition, greenhousesapv1alpha1.SCIMConfigNotProvidedReason, "SCIM config is not provided")
```

We need to avoid this. The auth type must come from spec and should not be guessed. This way we can switch case between multiple auth types and plus you won't need any validating webhooks.

In order to not break our apis you can do for example in organization_types -

```golang
type SCIMConfig struct {
	// URL to the SCIM server.
	BaseURL string `json:"baseURL"`

	// AuthType is the type of authentication to be used.
        // +kubebuilder:validation:Enum=basic;token
	// +kubebuilder:default="basic"
	AuthType string `json:"authType,omitempty"` // You can also make the string as type ScimAuthType string

	// User to be used for basic authentication.
	BasicAuthUser ValueFromSource `json:"basicAuthUser,omitempty"`

	// Password to be used for basic authentication.
	BasicAuthPw ValueFromSource `json:"basicAuthPw,omitempty"`
        .....
}
```
now you can switch case the authType and error out if the secrets for a particular auth type is not found. For our existing resources in the cluster they will all get basic by default since we supported only basic auth.

---

Done